src/core/Elsa.Core/Activities/Workflows/RunWorkflow/RunWorkflow.cs

```diff
@@ -98,22 +104,94 @@ public string ChildWorkflowInstanceId
+            //when retrying this activity if faulted, if there is only one with this activity id in the workflow, we dont need to use the hash because
+            //ChildWorkflowInstanceId will have the id of the subworkflow but, if it has failed in a loop, as ChildWorkflowInstanceId is metadata, it will always
+            //have stored just the last workflowinstance executed, but not the rest, so we need to discover what is the real workflowinstace using hashed input
+
+            string hash = default!;
 
-            if (workflowBlueprint == null || workflowBlueprint.Id == context.WorkflowInstance.DefinitionId)
-                return Outcome("Not Found");
+            if (RetryFailedActivities)
+            {
+                AlreadyExecutedChildren ??= new Dictionary<string, string>();
+                hash = HashHelper.Hash(context.Input);
+            }
 
-            var result = await _startsWorkflow.StartWorkflowAsync(workflowBlueprint!, TenantId, new WorkflowInput(Input), CorrelationId, ContextId, cancellationToken: cancellationToken);
-            var childWorkflowInstance = result.WorkflowInstance!;
-            var childWorkflowStatus = childWorkflowInstance.WorkflowStatus;
-            ChildWorkflowInstanceId = childWorkflowInstance.Id;
+            //We know it is a retry if ChildWorkflowInstanceId has a value here, but that value is not the real associated ChildWorkflow
+            if (RetryFailedActivities && !string.IsNullOrEmpty(ChildWorkflowInstanceId) && AlreadyExecutedChildren.ContainsKey(hash))
+            {
+                ChildWorkflowInstanceId = AlreadyExecutedChildren.GetValueOrDefault(hash);
+                childWorkflowInstance = await _workflowInstanceStore.FindByIdAsync(ChildWorkflowInstanceId);
 
-            context.JournalData.Add("Workflow Blueprint ID", workflowBlueprint.Id);
-            context.JournalData.Add("Workflow Instance ID", childWorkflowInstance.Id);
-            context.JournalData.Add("Workflow Instance Status", childWorkflowInstance.WorkflowStatus);
+                if (childWorkflowInstance == null)
+                {
+                    throw new Exception();
+                }
+                switch (childWorkflowInstance.WorkflowStatus)
+                {
+                    case WorkflowStatus.Idle:
+                    case WorkflowStatus.Finished:
+                    case WorkflowStatus.Suspended:
+                    case WorkflowStatus.Running:
+                    case WorkflowStatus.Cancelled:
+                        break;
+                    case WorkflowStatus.Faulted:
+                        await _workflowReviver.ReviveAndRunAsync(childWorkflowInstance, cancellationToken);
+                        break;
+                    default:
+                        break;
+                }
+                childWorkflowStatus = childWorkflowInstance.WorkflowStatus;
+                ChildWorkflowInstanceId = childWorkflowInstance.Id;
+
+                
```

Comment:
I noticed a few excessive blank lines here and there - let's tidy it up 😄

Reply:
My bad 😉 my brain was struggling hard with this Activity. Cleaning it next commit

Reply:
I can relate 😄